src/Interpreters/ExpressionAnalyzer.cpp

```diff
@@ -831,6 +974,49 @@ void SelectQueryExpressionAnalyzer::appendAggregateFunctionsArguments(Expression
                 getRootActions(argument, only_types, step.actions());
 }
 
+void SelectQueryExpressionAnalyzer::appendWindowFunctionsArguments(
+    ExpressionActionsChain & chain, bool /* only_types */)
+{
+    ExpressionActionsChain::Step & step = chain.lastStep(aggregated_columns);
+
+    for (const auto & f : window_function_descriptions)
+    {
+        // Requiring a constant reference to a shared pointer to non-const AST
+        // doesn't really look sane, but the visitor does indeed require it.
+        getRootActionsNoMakeSet(f.function_node->clone(), true, step.actions());
+
+        // Add empty INPUT with window function name.
+        // It is an aggregate function, so it won't be added by getRootActions.
+        ColumnWithTypeAndName col;
+        col.type = std::make_shared<DataTypeInt64>();
+        col.column = col.type->createColumn();
+        col.name = f.column_name;
+
+        step.actions()->addInput(col);
```

Comment:
Hm, it is a little bit strange to add window function name as input.
I think it would be logical to add a new action, after the calculation of window function arguments step, which will have this input.

Also it is not clear why we add window function column name to the list of required columns. I suppose it is possible that window function name is not actually used, and we can skip window function evaluation?

Reply:
Also, it probably may be done the same way as with aggregation.
The idea is that we can evaluate the list of columns after aggregation (`aggregated_columns`), and start building new `ExpressionActionsChain` starting from this list. Maybe we can have the same fixed list of columns after window function.

I don't say if it is a good approach and will be easier, though.

Reply:
Added a comment about the various approaches.

The aggregation one is logically similar to INPUT columns, but with more code. And we'll have to remove it anyway and add a proper WINDOW action. So I think we can keep the INPUT columns for now.